### PR TITLE
Repairing data capture anomalies

### PR DESCRIPTION
**Pull request by wendellpiez**
This addresses #50.

Also, it repairs data capture where they were perturbed by anomalies in AC-4(25) and SA-4(3)

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/oscal-content/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/oscal-content/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?

### CODE REVIEW

**Comment by wendellpiez** [comment]
@david-waltermire-nist @iMichaela

FWIW, my recommendation is to consider the current PR as a *stopgap* that improves over the current (garbled) data (with respect to SA-4(3) in particular but also AC-4(25)), without putting to bed all the (modeling/representation) issues raised here.

I *also recommend* bringing this to the FISMA team in the context of parameters (CPVs) in general. This is a good case of two parameters with a logical dependency we have not seen.

AC-4(25) should not be controversial: the PR corrects a typo. SA-4(3) however is more interesting.

For reference, here's a screen shot of SA-4(3) in the PDF:

image

The problem here is that while elsewhere, we have labeled parameters (assignments) referenced from inside selections, we do not elsewhere have - and have not stipulated processing logic for - the reverse, namely a selection (or another value) being referenced from a -- value? a label?

In illustration, here is an alternative to the variant given in the PR. (On the fence as to which is better.)

```xml
<param id="sa-4.3_prm_2">
    <label>organization-defined <insert param-id="sa-4.3_prm_2"/> methods</label>
</param>
<param id="sa-4.3_prm_3" depends-on="sa-4.3_prm_2">
    <select how-many="one or more">
        <choice>systems security</choice>
        <choice>privacy</choice>
    </select>
</param>
<part name="statement" id="sa-4.3_smt">
    <p>Require the developer of the system, system component, or system service to demonstrate
        the use of a system development life cycle process that includes:</p>
    <part name="item" id="sa-4.3_smt.a">...</part>
    <part name="item" id="sa-4.3_smt.b">
        <prop name="label">(b)</prop>
        <p><insert param-id="sa-4.3_prm_2"/>; and</p>
    </part>
    <part name="item" id="sa-4.3_smt.c">...</part>
</part>
```

Decomposing the logic here, why are there two nested parameters?

The requirement is evidently that both (a) the organization designate which methods (privacy or security or both) are appropriate, and (b) that the methods be described (to satisfy the CPV).

In other words, two separate questions are to be addressed, so this could also be

> **(b) [Assignment: organization-defined engineering methods] for [Selection (one or more): systems security; privacy] as appropriate;**

This would probably be easier to cross-check at the receiving end. (Is the selection correct? Are the methods described appropriate for the selection?)

In other words there are actually several options here especially if the FISMA team is willing to take up the question of parameter modeling (which they might be).

**Comment by iMichaela** [comment]
@wendellpiez:  I perfectly understand the problem and its complexity. I am not saying it is not a good solution, I only mentioned I want to review it more.
So, to sum, the parameters of SA-4(3).b. The `sa-4.3_prm_2` serves like an alias for the `sa-4.3_prm_3` alone, `sa-4.3_prm_4` alone OR `sa-4.3_prm_3` & (as in reunion) `sa-4.3_prm_4` together.  This is accurately reflected in:

```
<param id="sa-4.3_prm_2">
          <select how-many="one or more">
            <choice><insert param-id="sa-4.3_prm_3"/></choice>
            <choice><insert param-id="sa-4.3_prm_4"/></choice>
          </select>
        </param>
```

then the parameter is called in the control's statement.
```
<part name="item" id="sa-4.3_smt.b">
        <prop name="label">(b)</prop>
        <p><insert param-id="sa-4.3_prm_2"/>; and</p>
    </part>
```
When `sa-4.3_prm_2` is inserted in the statement (see below), the declaration above will be processed and the OSCAL SSP/component tool appears to have sufficient information to process and treat `sa-4.3_prm_2` as an alias. But how is the traceability maintain at the higher level and what is the SSP/component tool suppose to do with the `sa-4.3_prm_2` : replace it with:
`sa-4.3_prm_3` alone, 
`sa-4.3_prm_4` alone OR 
`sa-4.3_prm_3` & (as in reunion) `sa-4.3_prm_4` together as a list?

The declaration of 
```
<param id="sa-4.3_prm_3" depends-on="sa-4.3_prm_2">
```
and 
```
<param id="sa-4.3_prm_4" depends-on="sa-4.3_prm_2">
```

in the catalog is locally sufficient, or maybe in the profile too. But the traceability up to assessment layer and back down to `sa-4.3_prm_2`  which is the one appearing in the control's statement is the one I wanted to review with @david-waltermire-nist and you.

**Review by iMichaela — DISMISSED** [review]
The PR's changes are great. I have only the concerns expressed in the comments around SA-4(3).b parameters

**Review by GaryGapinski — COMMENTED** [review]
The commit message specifies SA-4(b) but the alteration occurred at control SA-4(3) (id sa-4.3 containing id sa-4.3_smt.b).

Also, the change results in something not the same as the intent (for better or worse) of the SP 800-53 authors to indicate that a selection is to be made in the context of an assignment, or in other words "assign the chosen selection(s)": manifest as two distinct ODPs in the origin document. Perhaps the OSCAL model should accommodate such "nested" parameters as they were defined and consider parameter reification (namely, what occurs subsequent to organizational parameter tailoring) if that is expected or required by a separate (i.e., not catalog) layer of OSCAL. Apologies if this is a naïve assessment, but while this change is an "interim" measure it seems inadequate. It does not define all constraints on catalog authorship which would avoid the perceived problems. In particular, one cannot render the control as it appears in SP 800-53r5 with 100% fidelity using the XML as input.

If desired, I can provide some diabolical examples (in XML) of what control catalog authors might innocently craft using just control statements with <insert>ed ODPs. While the OSCAL Catalog Layer can be considered a domain-specific language and is constrained by a schema, it is similar to English with control statements in natural language and arbitrary fragments replaced with <insert>s of <param>s, and the <insert>s instantiate into natural language fragments upon tailoring. Even contemporary (pre-OSCAL) ODP tailoring often results in bad English (and thus bad definitions of assessability). As long as English (or other natural languages) are the substance of control statements, puzzles will arise. At a minimum, control statement structure is not constrained by control parameters.

That problem might be diminished by prohibiting nested ODPs but I am unsure if that would constrain natural language in a manner considered unacceptably awkward. If it did, then it might be best to favor a (modified) DSL, rather than natural language, that enforces the desires of those vexed by their encounter with control SA-4(3).

That is tantamount to asserting that the essence of a control can only be formally and completely stated in an artificial rather than natural language and thus accommodating the natural language is not essential.

Is this nested (assertion of a selection) statement considered problematic simply because it requires an assessment (automated or not) to determine the correct observed value of two conjoined parameters?

**Comment by wendellpiez** [comment]
@GaryGapinski thanks so much for feedback here.

One reason we are preferring the interim solution here (in which both parameters are collapsed into one, and the nested one is taken to be implicit) is that (unlike the proposed solution it replaces), it is capable of supporting renditional fidelity to the control-as-printed, at least in a rude way with certain assumptions (i.e. the bracket convention for the notation, etc.) The refactored version it replaces didn't have that.

This is very much a stopgap solution and all your more general points are well taken. FWIW (with respect to larger issues you raise) my own feeling is that we need to accommodate *both* present-day awkward brittle natural-language substitutions (what I see as a 'mad libs') - which (at least up to a certain level of scale) can actually be a decent way to manage this kind of thing *if well governed* and well understood by users (big if) - and also, over time, a more rigorous approach supported by a DSL for expressing constraints and operational dependencies.

We do know we need to open the box of the parameter model indeed as you know the question has also been raised where parameters should be permitted to appear (beyond controls and catalogs). The example of SA-4(3)b (if I have the number right this time) should come up as one of the several examples that present questions of logical nesting or dependency chaining if you like. One alternative indeed would be to represent the expansion point for the second parameter inside the `param/label`, but we didn't wish to do this because it problematizes our present concept of what is a label (as opposed to a default or placeholder value or setting). Another solution could be to make explicit how values can be overridden at higher layers, and put it into the value (`value`) not the label.

We do have plenty of other cases of nesting, but elsewhere it is always an Assignment inside a Selection, not a Selection inside an Assignment. I agree that abandoning support for representations of constraints and dependencies in natural language is untenable. If nothing else this means that human operators must be keenly aware of the limits of automatability.

**Review by iMichaela — APPROVED** [review]

**Comment by iMichaela** [comment]
@wendellpiez Latest changes (replacement of `incorporated-into` with `moved-to`) looks good to me.

**Comment by wendellpiez** [comment]
FISMA team has now (March 2) accepted the version in this PR as superseding the older version now in the repo.

This PR should still be accepted at least until we have an updated feed from them (or another next/better version).

**Review by david-waltermire-nist — APPROVED** [review]
LGTM!
